File: 215/c/src/main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void printNums(int *nums, int numsSize, int k) {
  for (int i = 0; i < numsSize; i++) {
    printf("%d ", nums[i]);
  }
  printf("\n");
}

void swap(int *x, int *y) {
  int tmp = *x;
  *x = *y;
  *y = tmp;
}

int partition(int *nums, int low, int high) {
  int pivot = nums[high];
  int i = (low - 1);

  for (int j = low; j <= high - 1; j++) {
    if (nums[j] <= pivot) {
      i++;
      swap(&nums[i], &nums[j]);
    }
  }
  swap(&nums[i + 1], &nums[high]);
  return (i + 1);
}

int quickSellect(int *nums, int low, int high, int pos) {
  if (low < high) {
    int pivotIndex = partition(nums, low, high);
    if (pivotIndex == pos) {
      return nums[pivotIndex];
    } else if (pivotIndex > pos) {
      return quickSellect(nums, low, pivotIndex - 1, pos);
    } else {
      return quickSellect(nums, pivotIndex + 1, high, pos);
    }
  } else if (low == pos) {
    return nums[low];
  }
  return -1;
}

void quicksort(int *nums, int low, int high) {
  if (low < high) {
    int pivotIndex = partition(nums, low, high);
    quicksort(nums, low, pivotIndex - 1);
    quicksort(nums, pivotIndex + 1, high);
  }
}

int findKthLargest(int *nums, int numsSize, int k) {
  printNums(nums, numsSize, k);
  quicksort(nums, 0, numsSize - 1);
  // return quickSellect(nums, 0, numsSize - 1, numsSize - k);
  printNums(nums, numsSize, k);
  return nums[numsSize - k];
}

int main() {
  srand(time(NULL));
  int size = 5;
  int *numbers = (int *)malloc(size * sizeof(int));

  for (int i = 0; i < size; i++) {
    numbers[i] = 2;
  }

  numbers[0] = 12;
  numbers[1] = -12;
  numbers[2] = -2;
  numbers[3] = 122;
  numbers[4] = 2;

  int res = findKthLargest(numbers, size, 3);
  printf("%d ", res);
  return 0;
}

File: 215/rust/main/src/main.rs
fn partition(nums: &mut [i32], low: usize, high: usize) -> usize {
    let pivot = nums[high];
    let mut i = low as isize - 1;

    for j in low..high {
        if nums[j] <= pivot {
            i += 1;
            nums.swap(i as usize, j);
        }
    }
    nums.swap((i + 1) as usize, high);
    (i + 1) as usize
}

fn quick_select(nums: &mut [i32], low: usize, high: usize, pos: usize) -> i32 {
    if low < high {
        let pivot_index = partition(nums, low, high);
        if pivot_index == pos {
            return nums[pivot_index];
        } else if pivot_index > pos {
            return quick_select(nums, low, pivot_index - 1, pos);
        } else {
            return quick_select(nums, pivot_index + 1, high, pos);
        }
    } else if low == pos {
        return nums[low];
    }
    -1
}

fn find_kth_largest(nums: &mut [i32], k: usize) -> i32 {
    quick_select(nums, 0, nums.len() - 1, nums.len() - k)
}

fn main() {
    let size = 5;
    let mut numbers = vec![2; size];

    numbers[0] = 12;
    numbers[1] = -12;
    numbers[2] = -2;
    numbers[3] = 122;
    numbers[4] = 2;

    let res = find_kth_largest(&mut numbers, 3);
    println!("{}", res);
}

